add DNS correlation to connect tracking

nabilschear opened the pull request:
often i'd like to be able to track DNS domains that were associated with an IP at the time of connect.  So, I wired up a simple and hopefully low-overhead kprobe to udp_recvmsg that will capture DNS response packets and put them into a cache dictionary in user space.  When connect syscalls come in, then it will look for a corresponding DNS query in the cache.  It will also note the time delta between when the DNS response was received and when the connect happened.

This requires dnslib and expiringdict python packages.  dnslib is already required by an example in the bcc tree: https://github.com/iovisor/bcc/blob/b0bf04ac4042a6c004b15dcc5e40e12ae78020f9/examples/networking/dns_matching/dns_matching.py but I couldn't find anywhere to add these dependencies for a core tool.  Looking for help here.

yonghong-song commented:
[buildbot, test this please]

yonghong-song commented:
[buildbot, test this please]

brendangregg commented on tools/tcpconnect.py:
```diff
@@ -27,11 +28,15 @@
 examples = """examples:
     ./tcpconnect           # trace all TCP connect()s
     ./tcpconnect -t        # include timestamps
+    ./tcpconnect -d        # include DNS queries associated with connects
```
Hmm. I try to keep these tools simple -- do one job and do it well -- and this is adding UDP tracing to a "tcpconnect" tool, which seems like a violation. But it'll also be a common real world use case, so maybe it is ok in this case.

brendangregg commented on tools/tcpconnect.py:
```diff
@@ -224,7 +231,86 @@
+#
+# uses a percpu array of length 1 to store the dns_data_t off the stack to
+# allow for a maximum DNS packet length of 512 bytes.
+dns_bpf_text = """
```
Yes, this is good: making it its own text blob that we include or not include.

brendangregg commented on tools/tcpconnect.py:
```diff
@@ -224,7 +231,86 @@
+// single element per-cpu array to hold the current event off the stack
+BPF_PERCPU_ARRAY(dns_data,struct dns_data_t,1);
+
+int trace_udp_recvmsg(struct pt_regs* ctx)
```
What about TCP DNS? Isn't that a thing?

nabilschear replied:
yeah, theres' DNS over https and DNSSEC (where the max packet size is 4096).  This tool specifically handles classic DNS protocol only.

nabilschear replied:
clarified in readme

brendangregg commented on tools/tcpconnect.py:
```diff
@@ -224,7 +231,86 @@
+int trace_udp_recvmsg(struct pt_regs* ctx)
+{
+    __u64 pid_tgid = bpf_get_current_pid_tgid();
+    struct sock *sk = (struct sock*)PT_REGS_PARM1(ctx);
```
missing a space for kernel style: (struct sock *) ... and likewise elsewhere

nabilschear replied:
fixed, i think

brendangregg commented on tools/tcpconnect.py:
```diff
@@ -224,7 +231,86 @@
+{
+    __u64 pid_tgid = bpf_get_current_pid_tgid();
+    u32 zero = 0;
+    struct msghdr **msgpp = tbl_udp_msg_hdr.lookup(&pid_tgid);
```
does this always work (matching the PID of the DNS call to the PID of the TCP connect)?

What about systemd-resolved? Are there cases where an application asks a resolver to do name resolution on its behalf? If so, the PID of the DNS != PID of the TCP.

brendangregg replied:
Oh, this is for matching entry to ret of udp_recvmsg(), so disregard my comment.

brendangregg commented on tools/tcpconnect.py:
```diff
@@ -324,15 +492,21 @@ def depict_cnt(counts_tab, l3prot='ipv4'):
+    print("%-6s %-12s %-2s %-16s %-16s %-6s" % ("PID", "COMM", "IP", "SADDR",
+        "DADDR", "DPORT"), end="")
+    if args.dns:
+        print("%-10s %-20s" % ("DNS_DELTA", "DNS_QUERY"))
```
Other columns don't include "_". I'd rename these to "DNSus" (and don't include the "us" in the column rows) and "QUERY".

QUERY should also be "%s": it doesn't need the 20-char whitespace padding, it can grow/shrink as needed (indeed, I put such "%s"'s last so they don't unalign other columns).

nabilschear replied:
i updated it to DNS(ms) and explained it.  Fixed the query printing.

brendangregg commented on tools/tcpconnect_example.txt:
```diff
@@ -94,6 +111,8 @@ optional arguments:
   --cgroupmap CGROUPMAP
                         trace cgroups in this BPF map only
+  --mntnsmap MNTNSMAP   trace mount namespaces in this BPF map only
+  -d, --dns             include likely DNS query associated with each connect
```
please update the man page as well: man/man8/tcpconnect.8

nabilschear replied:
done

brendangregg reviewed (commented):
Thanks, some things to investigate are in this review.

brendangregg commented:
Ah sorry, just realized a higher-level change: I think this may be better suited as an addition to tcpconnlat, so you have both the TCP connection latency and DNS latency in one tool! Imagine:

```
# ./tcpconnlat -d
PID    COMM         IP SADDR            DADDR            DPORT LAT(ms) DNS(ms) QUERY
1201   wget         4  10.153.223.157   23.23.100.231    80    1.65    56.11   netflix.com
1201   wget         4  10.153.223.157   23.23.100.231    443   1.60    0.15    foo.com
1433   curl         4  10.153.223.157   104.20.25.153    80    0.75    1.52    localhost
[...]
```

nabilschear commented:
about tcpconlat, this tool doesn't actually show the DNS latency from request to response.  It is reporting the latency between the DNS response and a connect that uses IP address from the DNS response.  I was intending this latency to help you to determine how likely the DNS query's association with the connect is to be correct (as stated, it's not guaranteed).

nabilschear commented:
PS: i took a stab at trying to add the dnslib and cachetools dependencies to the right places (following the pattern set by python-netaddr), but I'm not totally sure I got that right.  I'll le the automated build tell me if it's happier.

yonghong-song commented:
[buildbot, test this please]

yonghong-song commented:
Tests failed with error:
```
30:   File "../../tools/tcpconnect.py", line 32, in <module>
30:     import dnslib
30: ImportError: No module named dnslib
```

yonghong-song commented:
maybe buildbot machine needs to add dnslib. @drzaeus77 ?

yonghong-song commented:
Could you directly add downloading dnslib and cachetools in scripts/build-rpm.sh and scripts/build-deb.sh these two scripts are called during buildbot testing. Hopefully this will solve dnslib and cachetools issue?

nabilschear commented:
> Could you directly add downloading dnslib and cachetools in scripts/build-rpm.sh and scripts/build-deb.sh these two scripts are called during buildbot testing. Hopefully this will solve dnslib and cachetools issue?

I just pushed a change to those scripts to install dnslib and cachetools

nabilschear commented:
[buildbot, test this please]

yonghong-song commented:
[buildbot, ok to test]

yonghong-song commented:
[buildbot, test this please]

yonghong-song commented:
On fedora:
```
+ sudo dnf install -y python-dnslib python-cachetools
Fedora 26 - x86_64 - Updates                     19 MB/s |  22 MB     00:01    
Fedora 26 - x86_64                               72 MB/s |  53 MB     00:00    
Last metadata expiration check: 0:00:05 ago on Tue 13 Oct 2020 12:56:23 AM UTC.
No match for argument: python-dnslib
Error: Unable to find a match
```

yonghong-song commented:
on ubuntu 16.04
```
+ sudo apt install -y python-dnslib python3-dnslib python-cachetools python3-cachetools

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

Reading package lists...
Building dependency tree...
Reading state information...
E: Unable to locate package python-dnslib
E: Unable to locate package python3-dnslib
```

on ubuntu 17.10
```
+ sudo apt install -y python-dnslib python3-dnslib python-cachetools python3-cachetools

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

Reading package lists...
Building dependency tree...
Reading state information...
The following package was automatically installed and is no longer required:
  grub-pc-bin
Use 'sudo apt autoremove' to remove it.
The following NEW packages will be installed:
  python-cachetools python-dnslib python3-cachetools python3-dnslib
0 upgraded, 4 newly installed, 0 to remove and 0 not upgraded.
Need to get 97.1 kB of archives.
After this operation, 524 kB of additional disk space will be used.
Err:1 http://us-west1.gce.archive.ubuntu.com/ubuntu artful/main amd64 python-cachetools all 1.1.6-1
  404  Not Found [IP: 104.198.11.147 80]
Get:2 http://us-west1.gce.archive.ubuntu.com/ubuntu artful/universe amd64 python-dnslib all 0.9.7+hg20170303-1 [39.3 kB]
Err:3 http://us-west1.gce.archive.ubuntu.com/ubuntu artful/universe amd64 python3-cachetools all 1.1.6-1
  404  Not Found [IP: 104.198.11.147 80]
Get:4 http://us-west1.gce.archive.ubuntu.com/ubuntu artful/universe amd64 python3-dnslib all 0.9.7+hg20170303-1 [39.4 kB]
Fetched 78.7 kB in 0s (149 kB/s)
E: Failed to fetch http://us-west1.gce.archive.ubuntu.com/ubuntu/pool/main/p/python-cachetools/python-cachetools_1.1.6-1_all.deb  404  Not Found [IP: 104.198.11.147 80]
E: Failed to fetch http://us-west1.gce.archive.ubuntu.com/ubuntu/pool/universe/p/python-cachetools/python3-cachetools_1.1.6-1_all.deb  404  Not Found [IP: 104.198.11.147 80]
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
```

ubunto 18.04 is successful
```
The following NEW packages will be installed:
  python-cachetools python-dnslib python3-cachetools python3-dnslib
0 upgraded, 4 newly installed, 0 to remove and 0 not upgraded.
Need to get 96.9 kB of archives.
After this operation, 524 kB of additional disk space will be used.
Get:1 http://us-west1.gce.archive.ubuntu.com/ubuntu bionic/main amd64 python-cachetools all 2.0.0-2 [9072 B]
Get:2 http://us-west1.gce.archive.ubuntu.com/ubuntu bionic/universe amd64 python-dnslib all 0.9.7+hg20170303-1 [39.3 kB]
Get:3 http://us-west1.gce.archive.ubuntu.com/ubuntu bionic/universe amd64 python3-cachetools all 2.0.0-2 [9136 B]
Get:4 http://us-west1.gce.archive.ubuntu.com/ubuntu bionic/universe amd64 python3-dnslib all 0.9.7+hg20170303-1 [39.4 kB]
debconf: unable to initialize frontend: Dialog
debconf: (Dialog frontend will not work on a dumb terminal, an emacs shell buffer, or without a controlling terminal.)
debconf: falling back to frontend: Readline
debconf: unable to initialize frontend: Readline
debconf: (This frontend requires a controlling tty.)
debconf: falling back to frontend: Teletype
dpkg-preconfigure: unable to re-open stdin: 
Fetched 96.9 kB in 0s (576 kB/s)
Selecting previously unselected package python-cachetools.
(Reading database ... 
(Reading database ... 5%
(Reading database ... 10%
(Reading database ... 15%
(Reading database ... 20%
(Reading database ... 25%
(Reading database ... 30%
(Reading database ... 35%
(Reading database ... 40%
(Reading database ... 45%
(Reading database ... 50%
(Reading database ... 55%
(Reading database ... 60%
(Reading database ... 65%
(Reading database ... 70%
(Reading database ... 75%
(Reading database ... 80%
(Reading database ... 85%
(Reading database ... 90%
(Reading database ... 95%
(Reading database ... 100%
(Reading database ... 79280 files and directories currently installed.)
Preparing to unpack .../python-cachetools_2.0.0-2_all.deb ...
Unpacking python-cachetools (2.0.0-2) ...
Selecting previously unselected package python-dnslib.
Preparing to unpack .../python-dnslib_0.9.7+hg20170303-1_all.deb ...
Unpacking python-dnslib (0.9.7+hg20170303-1) ...
Selecting previously unselected package python3-cachetools.
Preparing to unpack .../python3-cachetools_2.0.0-2_all.deb ...
Unpacking python3-cachetools (2.0.0-2) ...
Selecting previously unselected package python3-dnslib.
Preparing to unpack .../python3-dnslib_0.9.7+hg20170303-1_all.deb ...
Unpacking python3-dnslib (0.9.7+hg20170303-1) ...
Setting up python-dnslib (0.9.7+hg20170303-1) ...
Setting up python-cachetools (2.0.0-2) ...
Setting up python3-dnslib (0.9.7+hg20170303-1) ...
Setting up python3-cachetools (2.0.0-2) ...
```

nabilschear commented:
Thanks @yonghong-song for the help.  It doesn't look like dnslib is packaged on fedora and versions of ubuntu older than 18.04.  So, I moved the dependency into the code and documentation as optional.  Now, if you give it the option and you don't have the option, it will tell you so.  This is cheating a bit, but I think it's a reasonable approach to backward compatibility.  Hopefully this will fix build issues. Can you ask buildbot to start a build (it doesn't seem to listen to me =)?

brendangregg commented:
> ... reporting the latency between the DNS response and a connect that uses IP address from the DNS response

This is going to confuse everyone, who will assume it's DNS latency. I'd be very tempted to ditch the column and use some other indicator to pass this information on. E.g.:

```
# ./tcpconnlat -d
PID    COMM         IP SADDR            DADDR            DPORT LAT(ms) QUERY
1201   wget         4  10.153.223.157   23.23.100.231    80    1.65    netflix.com
1201   wget         4  10.153.223.157   23.23.100.231    443   1.60    foo.com(124ms)
1433   curl         4  10.153.223.157   104.20.25.153    80    0.75    localhost
[...]```

Then the tool has a threshold, say, 100ms, and adds the time after the query only if it exceeds the threshold. The man page explains that if a query string has a latency added, it's an indicator that the query string may not match the connection because that time elapsed between the lookup and the TCP connection.

nabilschear commented:
@yonghong-song how important is it to make the ubuntu 17.10 build work?  I can't install any packages in that environment because it's so out of date and not supported anymore.  I could use the same trick that I used for dnslib for cache tools too, but it seems unnecessary.  Thoughts?

yonghong-song commented:
ubuntu 17.10 failed.
```
The following NEW packages will be installed:
  python-cachetools python3-cachetools
0 upgraded, 2 newly installed, 0 to remove and 0 not upgraded.
Need to get 18.4 kB of archives.
After this operation, 100 kB of additional disk space will be used.
Err:1 http://us-west1.gce.archive.ubuntu.com/ubuntu artful/main amd64 python-cachetools all 1.1.6-1
  404  Not Found [IP: 35.203.149.125 80]
Err:2 http://us-west1.gce.archive.ubuntu.com/ubuntu artful/universe amd64 python3-cachetools all 1.1.6-1
  404  Not Found [IP: 35.203.149.125 80]
E: Failed to fetch http://us-west1.gce.archive.ubuntu.com/ubuntu/pool/main/p/python-cachetools/python-cachetools_1.1.6-1_all.deb  404  Not Found [IP: 35.203.149.125 80]
E: Failed to fetch http://us-west1.gce.archive.ubuntu.com/ubuntu/pool/universe/p/python-cachetools/python3-cachetools_1.1.6-1_all.deb  404  Not Found [IP: 35.203.149.125 80]
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
+ cleanup
+ [[ -d /tmp/debuild.BgnXW3 ]]
+ rm -rf /tmp/debuild.BgnXW3
Build step 'Conditional step (single)' marked build as failure
```

I agreed that ubuntu 17.10 is very old and it is okay to provide the feature as an option to only available say 18.04 and above.

nabilschear commented:
> > ... reporting the latency between the DNS response and a connect that uses IP address from the DNS response
> 
> This is going to confuse everyone, who will assume it's DNS latency. I'd be very tempted to ditch the column and use some other indicator to pass this information on. E.g.:
> 
> ```
> # ./tcpconnlat -d
> PID    COMM         IP SADDR            DADDR            DPORT LAT(ms) QUERY
> 1201   wget         4  10.153.223.157   23.23.100.231    80    1.65    netflix.com
> 1201   wget         4  10.153.223.157   23.23.100.231    443   1.60    foo.com(124ms)
> 1433   curl         4  10.153.223.157   104.20.25.153    80    0.75    localhost
> [...]
> ```
> 
> Then the tool has a threshold, say, 100ms, and adds the time after the query only if it exceeds the threshold. The man page explains that if a query string has a latency added, it's an indicator that the query string may not match the connection because that time elapsed between the lookup and the TCP connection.

great suggestion @brendangregg .  i implemented this and updated the man page.  Now that there's no latency related stuff by default, do you think that this is right to leave in tcpconnect rather than tcpconnlat?

yonghong-song commented:
@brendangregg could you check whether the output format is what you want or not?

nabilschear commented:
ping for @brendangregg, thoughts on output format and tcpconnect vs tcpconnlat?

brendangregg commented:
Sorry for the delay, I now have two broken systems where bcc won't compile.

On the third it compiled, but this message wasn't quite right:

```# ./tcpconnect.py -d
Error: The python packages dnslib and cachetools are required to use the -d/--dns option.
Install this package with:
	$ pip3 install dnslib cachetools
   or
	$ sudo apt-get install python3-dnslib python3-cachetools (on Ubuntu 18.04+)
```

/usr/bin/python is still Python 2 on our systems. Somehow. (I thought we EOL'd it.) I actually had to do a `sudo apt-get install python-dnslib python-cachetools`. Although I thought we weren't supposed to care about Python 2, so I'm not going to ask that the code say this.

Finally, I can see it working, and it's pretty useful! Thanks @nabilschear!

I'm ok to merge.

dalehamel commented:
I was trying to figure out why 0.17.0 never got pushed, it looks like it is because this broke the release process https://github.com/iovisor/bcc/runs/1361557409?check_suite_focus=true

I have #3162 up to avoid installing this packages which do not exist on the older ubuntu.
